scim2-sdk-common/src/main/java/com/unboundid/scim2/common/messages/PatchOperation.java

```diff
@@ -153,11 +146,201 @@ public <T> List<T> getValues(final Class<T> cls)
     {
-      JsonUtils.addValue(getPath() == null ? Path.root() :
-          getPath(), node, value);
+      Path path = (getPath() == null) ? Path.root() : getPath();
+      if (hasValueFilter(path))
+      {
+        validateAddOpWithFilter(path, value);
+        applyAddWithValueFilter(path, node, value);
+      }
+      else
+      {
+        JsonUtils.addValue(path, node, value);
+      }
+
       addMissingSchemaUrns(node);
     }
 
+    /**
+     * Indicates whether the provided attribute path has a value filter in the
+     * first element.
+     */
+    private boolean hasValueFilter(@Nullable final Path path)
+    {
+      return path != null
+          && path.size() > 0
+          && path.getElement(0) != null
+          && path.getElement(0).getValueFilter() != null;
+    }
+
+    /**
+     * Validates an add operation with a value selection filter. For more
+     * information, see {@link #applyAddWithValueFilter}.
+     * <p>
+     * This method imposes the following constraints:
+     * <ul>
+     *   <li> The value filter must be in the first element in the path.
+     *   <li> The value filter must be an {@link EqualFilter}. Paths such as
+     *        {@code addresses[type ne "home"]} are ambiguous and do not specify
+     *        the value that should be assigned to the {@code type} field.
+     * </ul>
+     *
+     * @param path The attribute path.
+     * @throws BadRequestException  If the value selection filter was in an
+     *                              invalid form for the add operation.
+     */
+    private void validateAddOpWithFilter(@Nullable final Path path,
+                                         @NotNull final JsonNode value)
+        throws BadRequestException
+    {
+      Filter filter;
+      if (path == null || !path.iterator().hasNext())
+      {
+        throw BadRequestException.invalidSyntax(
+            "The patch add operation was expected to contain a non-empty path");
+      }
+      if (value.isArray())
+      {
+        throw BadRequestException.invalidSyntax(
+            "Patch add operations with a filter cannot set the 'value' field to"
+                + " an array");
+      }
+
+      Iterator<Path.Element> it = path.iterator();
+      Path.Element firstElement = it.next();
+      filter = firstElement.getValueFilter();
+
+      FilterType filterType = (filter == null) ? null : filter.getFilterType();
+      if (filterType == null)
+      {
+        throw BadRequestException.invalidPath(
+            "The add operation contained an empty filter"
+        );
+      }
+      if (filterType != FilterType.EQUAL)
+      {
+        throw BadRequestException.invalidPath(String.format(
+            "The add operation contained a value selection filter of type '%s',"
+                + " which is not an equality filter",
+            filterType));
+      }
+      if (!it.hasNext())
+      {
+        // A Path with a filter must have a second element to specify the
```

Comment:
I think it's worth including this criteria in the method comment as well.

Reply:
Good idea. I've added this.